rewrite buffer implementation to eliminate evbuffer dependency

brian-pane opened the pull request:
*Description*:
Replace the use of libevent's evbuffer in `Buffer::OwnedImpl` with a new buffer implementation.

*Risk Level*: high
*Testing*: I ran the unit tests
*Docs Changes*: N/A
*Release Notes*: N/A
*Related Issues*: #4952 

Signed-off-by: Brian Pane <brianp+github@brianp.net>

mattklein123 commented:
Thanks @brian-pane. @envoyproxy/maintainers any takers on reviewing this? @jmarantz this seems like something you might enjoy. :wink:

jmarantz commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -12,99 +13,317 @@
-              "RawSlice != evbuffer_iovec");
-static_assert(offsetof(RawSlice, len_) == offsetof(evbuffer_iovec, iov_len),
-              "RawSlice != evbuffer_iovec");
+static uint64_t SliceSize(uint64_t data_size) {
```
If you make this a static method of OwnedBufferSlice then you can unit-test it. Also it should be sliceSize() whether a static method or static function.

jmarantz replied:
ping

jmarantz commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -12,99 +13,317 @@
-      buffer_.get(), fragment.data(), fragment.size(),
-      [](const void*, size_t, void* arg) { static_cast<BufferFragment*>(arg)->done(); }, &fragment);
+  OwnedBufferSlice(const void* data, uint64_t size) : OwnedBufferSlice(size) {
+    memcpy(&(base_[0]), data, size);
```
base_.get() ?

jmarantz commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -12,99 +13,317 @@
+
+  void* data() override { return &(base_[data_]); }
+
+  uint64_t dataSize() const override { return reserved_ - data_; }
```
you can omit the newlines between these one-liners.

jmarantz commented:
I need to dig into this but my first thought is that I'd like to see a perf test with this PR. My pattern lately has been to write the perf-test in a separate branch from master, and see how the current implementation performs. Then you can patch the perf test into this branch and measure any improvement.

I'm actually not sure if your motivation is to improve performance or simply to cut a troublesome dependency (and why is it troublesome?) but even if it's the latter, we should quantify the performance impact. In any case you should put the motivation for this PR in the description.

One such perf test is https://github.com/envoyproxy/envoy/blob/master/test/common/stats/thread_local_store_speed_test.cc

lizan commented on source/common/buffer/buffer_impl.h:
```diff
@@ -1,16 +1,98 @@
+ */
+class BufferSlice {
+public:
+  using Reservation = std::pair<void*, uint64_t>;
```
Instead of std::pair, use absl::string_view? It is more readable than `.first`/`.second` and comes with some utility function.

lizan replied:
Ah I think you're right string_view is const pointer. Then perhaps a simple struct with non-const pointer and size will work, I think in general we prefer named struct than std::pair.

jmarantz replied:
+1 on struct vs pair. Thus I suggest
```
struct Reservation {
  uint8_t* data_;
  uint64_t size_;
};
```
This may increase readability by reducing casts? See also RawSlice in include/envoy/buffer/buffer.h which is similar, but uses void* so will require casting for all uses.

brian-pane replied:
Maybe RawSlice itself should be the data type that represents a reservation. I'm thinking through the pros and cons now...

ggreenway commented:
Related: should we be calling `std::set_new_handler` (http://www.cplusplus.com/reference/new/set_new_handler/) with a function that calls `PANIC()` or `abort()` so that we get consistent results if a `std::bad_alloc` would have been thrown? Right now, the result will depend on which thread it occurs in, and whether there is a `try {}` block around it or not.

brian-pane commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -1,6 +1,7 @@
 #include "common/buffer/buffer_impl.h"
 
 #include <cstdint>
+#include <iostream>
```
That's left over from debugging. I'll post an update to remove it.

brian-pane commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -12,99 +13,319 @@
+  /**
+   * Compute a slice size big enough to hold a specified amount of data.
+   * @param data_size the minimum amount of data the slice must be able to store, in bytes.
+   * @ return a recommended slice size, in bytes.
```
Thanks for the catch on the the extraneous space; I'll fix that in my next update. With regard to listing the type after the `@return`, that style is apparently "outdated," based on the discussion here: https://github.com/envoyproxy/envoy/pull/2612#discussion_r168573142

jmarantz replied:
excluding the type makes sense to me; I hadn't noticed that thread.

brian-pane commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -12,99 +13,319 @@
+   * Offset in bytes from the start of the fragment to the start of the Data section
+   * (increases with each call to drain()).
+   */
+  uint64_t data_;
```
I made it `data_` for consistency with what the same concept is called in`OwnedBufferSlice`, but now that you mention it, `offset_` probably will be clearer in this situation.

jmarantz commented on test/common/ssl/ssl_socket_test.cc:
```diff
@@ -3536,11 +3536,14 @@ class SslReadBufferLimitTest : public SslSocketTest {
-        iovecs[0].len_ = 0;
-        iovecs[1].len_ = 0;
-        data.commit(iovecs, 2);
+        Buffer::RawSlice iovec1[1];
```
why does this test need to change?

brian-pane replied:
This test depended on an implementation detail of `evbuffer`: that particular call to `reserve` resulted in the creation of two separate slices inside the buffer. The new `OwnedImpl` implementation satisfies the reservation with a single slice in this case. One of the `TODO` items in `buffer_impl.cc` will probably result in the new implementation using two slices in this case in the future, to match how `evbuffer` works today, but I still wanted to make the test less dependent on implementation details.

jmarantz replied:
is that still true? or have you fixed this now?

brian-pane replied:
The original code here will now work once again, so I'll revert to that.

brian-pane commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -189,7 +469,7 @@ Api::SysCallIntResult OwnedImpl::write(int fd) {
 }
 
-OwnedImpl::OwnedImpl() : buffer_(evbuffer_new()) {}
+OwnedImpl::OwnedImpl() = default;
```
The `WatermarkBuffer` subclass depends on `OwnedImpl` having a default constructor.

brian-pane commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -148,21 +369,80 @@ Api::SysCallIntResult OwnedImpl::read(int fd, uint64_t max_length) {
-  evbuffer_ptr start_ptr;
-  if (-1 == evbuffer_ptr_set(buffer_.get(), &start_ptr, start, EVBUFFER_PTR_SET)) {
-    return -1;
+  // This implementation uses the same search algorithm as evbuffer_search(), a naive
```
I haven't seen this method show up in CPU profiles from production systems, so as far as I can tell it doesn't get used much in Envoy.

There are three calls to the method in Envoy currently, one of which appears to search untrusted inputs - and that one has a fixed pattern of length 12.

It's probably worth noting that, in the common case, both the current `evbuffer_search` implementation and this new code already benefit from an existing optimization: they use `memchr` to find the first character of the pattern, and for a large input size _n_, a typical vectorized `memchr` implementation only needs to do _n_/_v_ conditional branches, where _v_ is 16 or 32 depending on the size of the available SIMD registers.

jmarantz replied:
I was in error thinking this is easy to linearize, and I see the value in being able to do the vectorized memchr, so this looks like a good framework to start with until we see a perf problem.

But since the algorithm is very long with lots of corner-cases I think you need to add unit tests to cover a bunch of them. The current buffer_test.cc doesn't try to cover the underlying evbuffer_search impl which is assumed to be tested in the libevent repo, but now that you own it, you need to test it.

brian-pane commented on test/test_common/utility.cc:
```diff
@@ -96,23 +96,30 @@ bool TestUtility::buffersEqual(const Buffer::Instance& lhs, const Buffer::Instan
     return false;
   }
 
+  // Check whether the two buffers contain the same content. It is valid for the content
```
I originally used the `toString` approach, but it was slow enough to cause timeouts in some of the unit tests that compared really large buffers. :)

jmarantz commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -148,21 +364,81 @@ Api::SysCallIntResult OwnedImpl::read(int fd, uint64_t max_length) {
-                                        reinterpret_cast<evbuffer_iovec*>(iovecs), num_iovecs);
-  ASSERT(ret >= 1);
-  return ret;
+  // TODO(brian-pane) support the use of space in more than one slice.
```
It looks like this only works for num_iovecs==1? That seems like a problem if it's used in Envoy. If it's not, why not just remove the argument here and at the call-sites?

brian-pane replied:
It's more memory-efficient to support multiple iovecs, given the way Envoy does its reservations for socket reads. The only reason this PR doesn't support the multiple-iovec case yet is that the semantics are ill-defined. For example, there are cases where Envoy gets a reservation from an evbuffer with num_iovecs==2 and then calls commit on just the first iovec. It's unclear whether that is supposed to release the reservation on the second slice; I'll probably have to end up digging into the evbuffer implementation to find out the intended semantics. (I'm planning to do that in the very near term.)

jmarantz commented on test/test_common/utility.cc:
```diff
@@ -96,23 +96,30 @@ bool TestUtility::buffersEqual(const Buffer::Instance& lhs, const Buffer::Instan
+      }
+      auto lhs_str = static_cast<const uint8_t*>(lhs_slices[lhs_slice].mem_);
+      auto rhs_str = static_cast<const uint8_t*>(rhs_slices[rhs_slice].mem_);
+      if (lhs_str[lhs_offset] != rhs_str[rhs_offset]) {
```
If this is affecting test performance then I guess this regresses a bit; and you should probably use memcmp like it was before, using the std::min of the remaining lengths of the two fragments you are comparing.

I think you probably need to unit-test this test utility. I would also say that it should've been tested before too, but it's more complex now as you have taken on the larger burden of matching buffers that have different internal representations.

I'm also wondering whether this change in semantics affects any call-sites...I'm guessing no, but it's worth checking the call-sites (there are only ~10) and updating the doc in test/test_common/utility.h

brian-pane replied:
I'll add a unit test for this and update the comment in utility.h (which currently says "@return TRUE if the buffers are equal, false if not," so I'll clarify it to say that "equal" means that the linearizations of the content are equal)

jmarantz commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -12,99 +12,315 @@
 
-void OwnedImpl::add(absl::string_view data) {
-  evbuffer_add(buffer_.get(), data.data(), data.size());
+void OwnedImpl::add(const void* data, uint64_t size) { append(data, size); }
```
there's a comment in the base-class saying this method is deprecated; it's worth checking if there are remaining call-sites.

brian-pane replied:
I just checked, and there are some remaining call-sites within Envoy itself that depend on this method. `include/envoy/buffer/buffer.h` itself won't even compile with the deprecated method removed.

jmarantz commented on source/common/buffer/buffer_impl.h:
```diff
@@ -1,16 +1,98 @@
+   * Commit a Reservation that was previously obtained from a call to reserve().
+   * The Reservation's size is added to the Data section.
+   * @param reservation a reservation obtained from a previous call to reserve().
+   *        If the reservation is not from this BufferSlice, commit() will return false.
```
Under what circumstances would a user of this class commit a reservation from outside the BufferSlice? Just explain in the comment, especially in light of the comment above "the result of calling commit() out of order is undefined"

jmarantz commented on source/common/buffer/buffer_impl.h:
```diff
@@ -1,16 +1,98 @@
+   *       of order is undefined.
+   * @param size the number of bytes to reserve. The BufferSlice implementation MAY reserve
+   *        fewer bytes than requested (for example, if it doesn't have enough room in the
+   *        Reservable section to fulfill the whole request).
```
so it's the responsibility of the caller to check the reservation-size and create a new slice if needed?

brian-pane replied:
Almost -- It's the responsibility of the caller to check the reservation size and obtain the remaining from another slice if needed. That might mean creating a new slice, but it could also mean reusing some other existing slice (which `OwnedImpl` does when possible).

brian-pane commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -12,99 +13,319 @@
-      evbuffer_commit_space(buffer_.get(), reinterpret_cast<evbuffer_iovec*>(iovecs), num_iovecs);
-  ASSERT(rc == 0);
+  int64_t i = num_iovecs;
+  i--;
```
My rationale for doing the subtraction in signed space is that subtracting in unsigned space can cause an underflow (it's valid for `num_iovecs` to be zero). The only reason the overflow doesn't break anything is that the result is converted to a signed value of the same size (uint64_t to int64_t), so `i` ends up being `-1` as expected. But if somebody ever changes `num_iovecs` to, say, `unsigned` in the future, `i` will end up having a large positive value instead of `-1`. But I can at least replace:
```
int64_t i = num_iovecs;
i--;
```
with a cleaner equivalent:
```
int64_t i = static_cast<int64_t>(num_iovecs) - 1;
```

brian-pane commented:
I just added a couple of significant changes:
- Reservations from `BufferSlice` now use `RawSlice` rather than `std::pair`.
- In `RawSlice`, the pointer to the start of the data is now immutable (`void* const`), with some comments in `buffer.h` to explain why. The old implementation also depended on the caller not modifying `mem_` between the `reserve` and `commit` calls, so this change helps enforce pre-existing semantics.

Over the next couple of days, I plan to:
- Address the rest of the outstanding review comments.
- Add support for reservations that span more than one iovec (currently listed as a TODO comment in the PR, and important because otherwise this new implementation will use more memory than evbuffer).

jmarantz commented:
If you have something functional, but whose performance/memory usage has yet to be proven to be the same or better as libevent's impl, you might want to check it in as an alternate implementation that's tested but not in use. That way you can iterate on it, with more frequent and smaller PRs.

One approach toward this is to parameterize your test, as shown in PR #5414  https://github.com/envoyproxy/envoy/blob/22091159fd2655fc5c34f754f36ad3724a5ec392/test/common/stats/symbol_table_impl_test.cc#L31

Note that checking in a not-quite-good-enough impl will not go down well with folks that are deploying Envoy from HEAD every week, as opposed to waiting for a numbered release.

jmarantz commented:
OK...you sure you don't want to try to check this in as an alternate impl, so you can iterate in small PRs? I think it'd be easier to review, rather than waiting till performance perfection?

What's your specific issue with deque? As it happens, a while back on a different project, I had a measured performance issue with having lots of small deques and wound up re-implementing a subset of the std::deque API using a vector and getting a lot of performance benefit for my use-case: https://github.com/apache/incubator-pagespeed-mod/blob/master/pagespeed/kernel/base/vector_deque.h , with a speed-test at https://github.com/apache/incubator-pagespeed-mod/blob/master/pagespeed/kernel/util/deque_speed_test.cc

brian-pane commented:
Performance benchmark results:
```
------------------------------------------------------------------------
Benchmark                              Original           New
------------------------------------------------------------------------
BufferCreateEmpty                         80 ns         10 ns
BufferCreate/1                           165 ns        132 ns
BufferCreate/4096                        243 ns        220 ns
BufferCreate/16384                       596 ns        576 ns
BufferCreate/65536                      3685 ns       2501 ns
BufferAddSmallIncrement/1                 14 ns         10 ns
BufferAddSmallIncrement/2                 14 ns         10 ns
BufferAddSmallIncrement/3                 14 ns         10 ns
BufferAddSmallIncrement/4                 14 ns         10 ns
BufferAddSmallIncrement/5                 14 ns         10 ns
BufferAddString/1                         14 ns         10 ns
BufferAddString/4096                     244 ns        316 ns
BufferAddString/16384                    789 ns        905 ns
BufferAddString/65536                   3885 ns       3465 ns
BufferAddBuffer/1                         36 ns         22 ns
BufferAddBuffer/4096                     261 ns        316 ns
BufferAddBuffer/16384                    817 ns        902 ns
BufferAddBuffer/65536                   3966 ns       3450 ns
BufferPrependString/1                     14 ns          9 ns
BufferPrependString/4096                 303 ns        310 ns
BufferPrependString/16384                880 ns        887 ns
BufferPrependString/65536               3604 ns       3566 ns
BufferPrependBuffer/1                    707 ns        108 ns
BufferPrependBuffer/4096                 197 ns        106 ns
BufferPrependBuffer/16384                197 ns        105 ns
BufferPrependBuffer/65536                198 ns        105 ns
BufferDrain/1                             41 ns         26 ns
BufferDrain/4096                         187 ns        215 ns
BufferDrain/16384                        594 ns        634 ns
BufferDrain/65536                       3702 ns       2816 ns
BufferDrainSmallIncrement/1               12 ns          5 ns
BufferDrainSmallIncrement/2               12 ns          5 ns
BufferDrainSmallIncrement/3               12 ns          5 ns
BufferDrainSmallIncrement/4               12 ns          5 ns
BufferDrainSmallIncrement/5               12 ns          5 ns
BufferMove/1                              47 ns         20 ns
BufferMove/4096                           47 ns         21 ns
BufferMove/16384                          47 ns         20 ns
BufferMove/65536                          47 ns         20 ns
BufferMovePartial/1                       62 ns         23 ns
BufferMovePartial/4096                119143 ns     126331 ns
BufferMovePartial/16384               481011 ns     310920 ns
BufferMovePartial/65536              1901842 ns    1084919 ns
BufferReserveCommit/1                    172 ns         12 ns
BufferReserveCommit/4096                 160 ns        153 ns
BufferReserveCommit/16384                159 ns        154 ns
BufferReserveCommit/65536                944 ns        146 ns
BufferReserveCommitPartial/1              24 ns         13 ns
BufferReserveCommitPartial/4096           27 ns         13 ns
BufferReserveCommitPartial/16384          27 ns         13 ns
BufferReserveCommitPartial/65536          27 ns         13 ns
BufferLinearizeSimple/1                   95 ns         89 ns
BufferLinearizeSimple/4096                96 ns         90 ns
BufferLinearizeSimple/16384               95 ns         89 ns
BufferLinearizeSimple/65536               95 ns         90 ns
BufferLinearizeGeneral/1                  98 ns         91 ns
BufferLinearizeGeneral/4096              729 ns        646 ns
BufferLinearizeGeneral/16384            2696 ns       2529 ns
BufferLinearizeGeneral/65536           11747 ns       9972 ns
BufferSearch/1                            30 ns         24 ns
BufferSearch/4096                        125 ns        121 ns
BufferSearch/16384                       382 ns        374 ns
BufferSearch/65536                      1775 ns       1757 ns
BufferSearchPartialMatch/1               471 ns        279 ns
BufferSearchPartialMatch/4096          41489 ns      24474 ns
BufferSearchPartialMatch/16384        164634 ns      96503 ns
BufferSearchPartialMatch/65536        657967 ns     386232 ns
```

brian-pane commented:
@jmarantz now that the performance of this is competitive with `evbuffer`, I think it's stable enough for further code review. Some notes on the latest changes:
- I ended up not doing reference-counted buffer slices, since the performance was okay without them. Adding them in the future could speed up the `BufferMovePartial` benchmark test, as well as various real use cases in Envoy that move N bytes from the start of one buffer to the end of another. For now, leaving out reference-counting kept the implementation a lot simpler.
- `Buffer::OwnedImpl` now caches the total length of the buffer, to avoid iterating through all the slices to sum their sizes in the `length()` method.
- There's a custom deque implementation that uses an inline ring buffer and then switches to a heap-allocated buffer if the number of slices grows too large.

htuch commented:
Just as a drive-by meta-comment, it's great that we now have microbenchmarks for the buffer implementations, it would be also nice to be able to point to which operations really are important on the dataplane fastpath, since different buffer implementations in the future will inevitably require tradeoffs to be made.

htuch commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -43,71 +56,213 @@ void OwnedImpl::add(const Instance& data) {
 
 void OwnedImpl::prepend(absl::string_view data) {
-  evbuffer_prepend(buffer_.get(), data.data(), data.size());
+  uint64_t size = data.size();
```
More drive-by; it would be super-awesome to have a buffer fuzzer. You could use https://github.com/envoyproxy/envoy/blob/master/test/common/http/header_map_impl_fuzz_test.cc as an example of how to write this kind of data structure fuzzer. The interesting thing here is you could easily state correctness conditions I think, so you might be able to do a stronger than "it doesn't crash" fuzzer. Fine if you want to create a TODO or file an issue, but I think this would increase confidence in implementation correctness.

jmarantz commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -43,71 +56,213 @@ void OwnedImpl::add(const Instance& data) {
+  // NOLINTNEXTLINE(bugprone-assert-side-effect)
+  ASSERT(length_ == [this]() -> uint64_t {
+    uint64_t length = 0;
+    for (size_t slice_index = 0; slice_index < slices_.size(); slice_index++) {
```
`for (const auto& slice : slices_) {`

brian-pane replied:
The "for (item : collection)" syntax only works if the collection class has iterator methods defined, and my custom deque class doesn't. :(

jmarantz commented on test/common/buffer/buffer_test.cc:
```diff
@@ -14,6 +14,252 @@ namespace Envoy {
+    ASSERT_EQ(initial_capacity - 10, slice->reservableSize());
+
+    // Verify that a reservation can only be committed once.
+    ASSERT_FALSE(slice->commit(reservation));
```
it looks like these stanzas have a lot of common code...could you consider making a test-class and putting a helper-method in there so there's fewer/smaller repeated blocks of code?

jmarantz commented on test/common/buffer/buffer_test.cc:
```diff
@@ -14,6 +14,252 @@ namespace Envoy {
+
+    // Commit the entire reserved size.
+    bool committed = slice->commit(reservation);
+    ASSERT(committed);
```
In general s/ASSERT/ASSERT_TRUE/g in the _test.cc file.  Those are very different, as the ASSERT_TRUE will stop the test, but not crash the process, and will allow other tests to run.

Moreover we use EXPECT_EQ, EXPECT_TRUE etc whenever it makes sense to keep on testing more cases even if the current statement fails.

brian-pane replied:
Thanks for the context, I'll switch to `EXPECT_*`.

lizan commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -12,26 +12,39 @@
+}
 
-void OwnedImpl::add(const void* data, uint64_t size) { evbuffer_add(buffer_.get(), data, size); }
+void OwnedImpl::add(const void* data, uint64_t size) {
```
Yes for mostly historical reasons, this exists before we have string_view in the codebase. prepend is introduced later so we only have string_view version.

brian-pane replied:
The problem I see with string_view is that it uses char* as its pointer type. The signedness of char is unspecified, so any client code that operates on binary protocols would have to reinterpret_cast between char* and uint8_t*. We'd end up with fewer static_casts inside the buffer impl, but at the cost of a lot of new reinterpret_casts throughout the filter and codec code.

brian-pane replied:
It might be useful to talk through the three cases of casts individually:

`reinterpret_cast`:
I just double-checked buffer_impl.cc, and all the reinterpret_casts are in the old implementation.

`static_cast`:
Most of the static_casts are there to turn void* provided by callers into uint8_t*. The equivalent code in the previous implementation uses C-style casts in the same spots: [libevent/buffer.c](https://github.com/libevent/libevent/blob/63c6669fd831088825dfc0000737a33853bb36e8/buffer.c). So my perspective has been that the static_casts make the new impl easier to read than the old impl.

The other uses of static_cast are in the move methods, to enforce the constraint that moves can only occur between the same buffer subclass. The old implementation does the same thing. I think your point about validating this with a runtime "ASSERT(dynamic_cast(..." is reasonable; I'll add that.

`const_cast`:
There are two const_casts in the new buffer implementation, and they're both there to support the change I made in `RawSlice`: replacing `void* mem_` with `void* const mem_`. That was a conscious tradeoff: making the field in the public struct `const` helps ensure that callers can't corrupt a reservation, since the reserve/commit model depends on the pointer not changing. The price of that external constness is a pair of const_casts inside the buffer impl. Alternatively, I could remove the `const` from the public `RawSlice` struct declaration and thus eliminate the need for the const_casts, but the tradeoff there would be less protection against callers corrupting the reservations. What do you think: is it cleaner to keep the const in the public interface (and incur the need for the const_casts internally), or to revert to the non-const public interface (and maybe just use a warning in the comments to persuade callers not to modify the `mem_` field)?

(If we could start over from scratch, the cleaner solution would be to replace `RawSlice` with a class containing private fields and public accessor methods. But doing that now would break people's third-party filter code.)

There used to be a lot more static_casts in the new implementation, but those went away with one of the prior revisions to this PR where I changed the internal slice API to use `uint8_t*` instead of `void*`.

lizan commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -43,71 +56,213 @@ void OwnedImpl::add(const Instance& data) {
+uint64_t OwnedImpl::length() const {
+  // When running in debug mode, verify that the precomputed length matches the sum
+  // of the lengths of the slices.
+  // NOLINTNEXTLINE(bugprone-assert-side-effect)
```
Why did you need this? Your ASSERT shouldn't have side effect, it might be possibly due to the for loop isn't const range loop.

brian-pane commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -43,71 +56,213 @@ void OwnedImpl::add(const Instance& data) {
+    size -= copy_size;
+    length_ += copy_size;
+    new_slice_needed = true;
+  }
```
Sure, I'll add a test to exercise the case where it goes through the loop twice.

brian-pane commented:
/retest

brian-pane commented:
Next I'm going to make a buffer fuzz test, per @htuch's suggestion. I'll do that as a separate PR.

brian-pane commented:
Update: I had to give up on the buffer fuzz test, due to problems with the associated Bazel tooling.

jmarantz commented on test/common/buffer/buffer_test.cc:
```diff
@@ -14,6 +14,260 @@ namespace Envoy {
+    slices.emplace_back(
+        std::make_unique<DummySlice>(slice1, [&slice1_deleted]() { slice1_deleted = true; }));
+    EXPECT_FALSE(slices.empty());
+    EXPECT_EQ(1, slices.size());
```
This particular one (and similar cases) should be an ASSERT_EQ because you want to exit this test-method if it fails, rather than crashing on the slices.front() below, thus allowing the other test-methods to still run.

jmarantz commented on test/common/buffer/buffer_test.cc:
```diff
@@ -14,6 +14,260 @@ namespace Envoy {
+    EXPECT_TRUE(sliceMatches(slices.back(), slice2));
+
+    // Prepend a view to the deque, to exercise the ring buffer wraparound case.
+    const std::string slice3 = "slice2";
```
did you mean to have "slice2" here? Not "slice3"? Not sure it matters.

alyssawilk commented:
I'm not sure this is totally orthogonal.  Given the structure of this class being used by several subclasses, I think what we'd do in-house is either have BufferImplOld and BufferImplNew exist side by side in the cc file with a shim wrapper class which creates the right one, or since it looks like strong .h overlap and there's roughly a dozen functions just have a boolean old_behavior_ and switch code paths based on the value of flag/run-time-variable at start-up.   It's ugly but hopefully transient ugliness and makes for safer deployment

brian-pane commented on include/envoy/buffer/buffer.h:
```diff
@@ -18,9 +18,21 @@ namespace Buffer {
  */
 struct RawSlice {
-  void* mem_ = nullptr;
+  void* const mem_ = nullptr;
```
> Actually I just looked over the Buffer interface and it's mostly char*.

What Buffer interface are you referencing? Buffer::Instance and Buffer::OwnedImpl don't appear to use char* in their APIs at all.

Switching from void* to char* in the public-facing interface of the Buffer classes would require a lot of `reinterpret_cast`s. The reason is that most of the protocols Envoy supports are binary, and they need to parse their messages as uint8_t rather than char because the unspecified (implementation-dependent) signedness of char makes it tricky to use safely.

Switching from void* to uint8_t* would still require `reinterpret_cast`s, but I think the amount of affected code would be smaller: just the filters and codecs that currently treat buffers as sequences of characters. Those tend to be legacy protocol implementations, like HTTP/1.

jmarantz replied:
Sorry about that; I missed the point that RawSlice is an exposed interface of a concrete struct, used at many call-sites in the codebase. My high level bit is to reduce const_cast in the impl because it requires scrutiny.

mattklein123 commented:
> I'm not sure this is totally orthogonal. Given the structure of this class being used by several subclasses, I think what we'd do in-house is either have BufferImplOld and BufferImplNew exist side by side in the cc file with a shim wrapper class which creates the right one, or since it looks like strong .h overlap and there's roughly a dozen functions just have a boolean old_behavior_ and switch code paths based on the value of flag/run-time-variable at start-up. It's ugly but hopefully transient ugliness and makes for safer deployment

Sorry I meant the question of how to plumb through a feature flag for this is somewhat (bot not entirely orthogonal). I'm agreeing with @alyssawilk that for this case, it's not reasonable to provide a `Runtime::Loader` reference to every place that constructs a buffer and we need a global solution. What that global solution looks like I think needs to be discussed. @alyssawilk is that fair?

alyssawilk commented:
Yep, totally fair.  For the sake of expediency I'm OK one-offing this with whatever hack-around we think is most expedient.  I'd prefer something command-line and/or runtime over a compile-time fix but whatever works.

mattklein123 commented:
> Yep, totally fair. For the sake of expediency I'm OK one-offing this with whatever hack-around we think is most expedient. I'd prefer something command-line and/or runtime over a compile-time fix but whatever works.

+1 on either CLI or runtime. I know we are having a meeting about this tomorrow and then will have an issue with public discussion, but FWIW I'm leaning towards making Runtime a static/global and an exception to our no static/globals rule, and then having different implementations (FS, gflags, etc.). I think this would provide the most flexibility and make everyone happy.

brian-pane commented:
> Maybe it could be a 0 or 100% switch only read once on program load?
Yes, as long as it's only read once at startup, everything should work fine.

brian-pane commented:
The outstanding errors on the circleci:format test are from the`check_spelling_pedantic` step, which flags "deque," "tuple," and "reservable" as invalid words.

jmarantz commented:
Add the extra words to tools/spelling_dictionary.txt in your PR.

jmarantz commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -12,99 +13,319 @@
+  }
+
+  void* data() override {
+    return static_cast<uint8_t*>(const_cast<void*>(fragment_.data())) + data_;
```
ping

brian-pane commented on source/common/buffer/buffer_impl.cc:
That's the old implementation, so I recommend not spending resources on refactoring it.

brian-pane replied:
I added an "ASSERT(dynamic_cast...)" check for both the old and new code paths.

brian-pane commented on test/common/buffer/buffer_test.cc:
```diff
@@ -14,6 +14,260 @@ namespace Envoy {
+    reservation.len_ = 5;
+    bool committed = slice->commit(reservation);
+    expectCommitSuccess(committed, *slice, 15, initial_capacity - 15);
+  }
```
I already refactored this to encapsulate the common logic in `expectReservationSuccess` and `expectCommitSuccess`

htuch commented:
@brian-pane I have some cycles and could work on a buffer fuzzer.. did you make any progress here or would it make sense for me to take this?

brian-pane commented:
@htuch I wasn't able to make any progress on a fuzzer, so I'll be grateful for any help.

brian-pane commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -12,188 +12,626 @@
+    ASSERT(data.length() == 0);
+    static_cast<LibEventInstance&>(data).postProcess();
+  } else {
+    OwnedImpl& other = static_cast<OwnedImpl&>(data);
```
I added an explanatory comment.

jmarantz commented on source/common/buffer/buffer_impl.h:
```diff
@@ -1,16 +1,434 @@
+ *       revealed that std::deque was too slow to reach performance parity with the
+ *       prior evbuffer-based buffer implementation.
+ */
+class SliceDeque {
```
this faster deque is great and is justified in comments. But I think it'd be better to move it into its own file outside of the buffer impl. and into its own header. Also github may be failing me but we need to test your new deque in isolation IMO.

My suggestion is to actually pull this into source/common/common/fast_deque.h or similar, add tests, and check that in as a separate PR. Should be easy to review and then shrink this one.

jmarantz replied:
also for tests you could start by borrowing these: https://github.com/apache/incubator-pagespeed-mod/blob/master/pagespeed/kernel/base/vector_deque_test.cc

brian-pane replied:
There's a set of tests for the new deque in `test/common/buffer/buffer_test.cc` in this PR.

brian-pane replied:
If we put this deque class in source/common/common it probably should be generic (templatized) rather than hard-coded to be a deque of slices specifically. Does the API for `SliceDeque` look reasonable as a starting point for a generic Deque class? If so, I'll convert it to a template and move it to source/common/common.

jmarantz replied:
Oh sorry I missed that you did not templatize it. If you can that'd be awesome, but that's unreasonable for me to ask in this PR. Splitting it out into its own PR seems reasonable, and putting it in its own file in the buffer directory would be good anyway so it can be individually reviewed, and easily templatized later if needed. I had missed that you added tests because ^F in github failed to find them in buffer_test.cc as it was too large to display by default, but in any case they and the class should probably be in their own file.

brian-pane commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -43,157 +61,556 @@ void OwnedImpl::add(const Instance& data) {
+      }
+      uint64_t data_size = slice->dataSize();
+      if (data_size <= start) {
+        start -= data_size;
```
This is a valid case. It happens when the caller does something like:
```
// Copy 5 bytes of data, starting at offset 10 from the start of the buffer
buffer.CopyOut(10, 5, destination);
```
and the buffer internally contains a slice holding 3 bytes of data followed by a slice containing 1KB. It's okay that the first slice contains fewer than 10 bytes; that just means the copy loop needs to skip past that slice. At that point, there are 7 bytes left to skip. Once the copy loop reaches a slice whose dataSize is > the number of bytes left to skip, it starts copying from that remaining offset in the slice. And after that slice, if there's any more copying needed, it starts from the very beginning of the next slice (i.e., there's no need to skip any more leading bytes), which the start=0 below ensures.

jmarantz replied:
Cool. Can you add comments? Maybe just point to the test-case that wakes the code-branch up, so it'll be terse.

brian-pane commented on source/server/options_impl.cc:
```diff
@@ -118,6 +118,9 @@ OptionsImpl::OptionsImpl(int argc, const char* const* argv,
       "", "enable-mutex-tracing", "Enable mutex contention tracing functionality", cmd, false);
 
+  TCLAP::SwitchArg use_libevent_buffer(
+      "", "use-libevent-buffers", "Use the original libevent buffer implementation", cmd, false);
```
Based on my experiences at an organization that deploys Envoy from master, I'd prefer to have the new implementation be enabled by default.

jmarantz replied:
OK I'll let senior committers decide. Here's two rationales for deploying binaries with new features off-by-default.
 * Say someone deploys this that doesn't know about this new impl. They might not even know to turn it off if something goes wrong, and will instead be forced to roll back. When we import master into our repo there are often ~100 PRs that come along with it, so it's easy to miss. But those who are aware, and turn it on, can confirm it's stable in production and then we can write an easy PR to flip the default.
 * This is likely a perf improvement, and you may want to track carefully what the impact is, e.g. to write about it in a self-eval for your company for example. But if this is committed coincident with a bunch of other changes your speed win may be lost in the noise of other changes.

What's the downside of old-impl-by-default?

Having said that I'm OK if @alyssawilk and @mattklein123 want to have this on by default.

mattklein123 replied:
This is so risky that I would personally be in favor of off by default, along with a clear announcement/release note, etc. for people to test.

brian-pane commented on source/common/buffer/buffer_impl.h:
```diff
@@ -1,16 +1,434 @@
+   */
+  static SlicePtr create(uint64_t capacity) {
+    uint64_t slice_capacity = sliceSize(capacity);
+    return SlicePtr(new (slice_capacity) OwnedSlice(slice_capacity));
```
This is the 2nd option in the "Syntax" section near the top of https://en.cppreference.com/w/cpp/language/new :
`::`_(optional)_ `new` `(`placement_params`)`_(optional)_ `type` initializer _(optional)_ (2)

As far as I know, they don't show parentheses around the "initializer" part because it's just one of the supported initializer forms (with curly-brace style initializers being the other).

(Interestingly, the first option in that section is the same thing except with parentheses wrapped around 
the type name, which I don't think I've seen used in practice yet.)

This section has a little more detail: https://en.cppreference.com/w/cpp/language/new#Placement_new

Now that you mention it, I have concerns about the idiom used in heap_stat_data.cc, as it explicitly uses `malloc` for the allocation but just lets the compiler use the standard `::delete` implementation for the deallocation. I can't find a reference that guarantees `::delete` will use `free` by default. And if anybody ever plugs in an override for the global `::new` and `::delete` that uses a different underlying API than `malloc`/`free`, that code could stop working.

jmarantz replied:
re HeapStatData the constructor and destructor are private; it needs to be freed by calling its class free() method which then calls ::free(). You are absolutely right that deleting a malloced pointer is not safe -- and valgrind  checks this -- but this code can't do that from outside the class because of the private ctor/dtor.

Thanks for the explanation about your use of placed new.

Basically the bottom line is that most modern C++ is insulated from this detail when we are trying to hyper-optimize we need to be more explicit with memory and indirection. To me, the malloc/free/private-dtor pattern is a little clearer -- but that's probably just because I'm used to it.

brian-pane replied:
Ah, thanks for the pointer; I missed the private destructor in HeapStatData. The advantage I see in the placement new idiom (compared to using `malloc`) is that it works safely with a public destructor, which in turn enables the use of `unique_ptr` and `shared_ptr`.

jmarantz replied:
Actually unique_ptr works fine with a custom free(), and is used for that class:

https://github.com/envoyproxy/envoy/blob/90a7a3eb8f477350a7175cc0d98487bbed1c4188/source/common/stats/heap_stat_data.cc#L23

brian-pane commented on source/common/buffer/buffer_impl.h:
```diff
@@ -1,16 +1,434 @@
+    /*
+    external_ring_ = std::make_unique<SlicePtr[]>(InlineRingCapacity);
+    ring_ = external_ring_.get();
+    */
```
Thanks for the catch; I'll fix that and post an update shortly.

brian-pane commented on source/common/buffer/buffer_impl.h:
```diff
@@ -1,16 +1,434 @@
+    ring_ = external_ring_.get();
+    */
+    ASSERT(nullptr == external_ring_);
+    ASSERT(ring_ == &(inline_ring_[0]));
```
Definitely, I'll remove these.

brian-pane commented on source/common/buffer/buffer_impl.cc:
Thanks for the suggestion. Fixed in the latest update that I just pushed.

brian-pane commented on test/common/buffer/owned_impl_test.cc:
```diff
@@ -186,6 +358,32 @@ TEST_F(OwnedImplTest, ToString) {
+TEST_P(OwnedImplTest, AppendSliceForTest) {
+  static constexpr size_t NumInputs = 3;
+  static constexpr const char* Inputs[] = {"one", "2",
+                                           ""
```
Fixed

jmarantz commented:
Oh and I let's be certain we're all on the same page for this being on-by-default.

As I recall, in the past it was useful for someone at Lyft to patch & manually canary potentially risky changes, though I think that tells you only about crashing and not about response to load, etc.

brian-pane commented:
/retest

brian-pane commented:
The new implementation now is disabled by default, and I added an entry to the release notes to explain how to turn enable it. I've addressed all the code review comments, so I think this is ready for final review.

jmarantz commented:
Can you go through all my past comments and mark them 'resolved' if they are? I still see a ton of them.

brian-pane commented:
@jmarantz to save some time, could you specify what your unaddressed concerns are? Thanks.

jmarantz commented:
I think the most efficient mechanism is to walk through all the github comments and, if they are resolved, click the "Resolve Conversation" button. We should be left then with only the unresolved ones which will be easier to see then.

Mostly I think my issues at this point are around commenting. I also am still slightly bothered by the operator-delete override just because it feels a little like it might not be future-proof or fully portable (especially given the symptom you described of not having it) and I think the explicit deleter override with unique_ptr would be more robust. But I don't feel strongly enough about that to insist.

brian-pane commented:
Overriding the new and delete operators for a class has been an idiom in C++ for a long time, in part because it works transparently with lots of different smart-pointer implementations (unique_ptr, shared_ptr, Boost's intrusive_ptr, the new C++20 std::atomic<shared_ptr> stuff, etc).

In contrast, the idiom used in HeapStatData (special wrapper functions for allocation and deallocation) requires extra glue code for each type of smart pointer that one wants to support. So for easier future maintenance, I'm a fan of overriding new and delete instead.

jmarantz commented on source/common/buffer/buffer_impl.h:
```diff
@@ -1,16 +1,427 @@
+
+private:
+  static void* operator new(size_t object_size, size_t data_size) {
+    return ::operator new(object_size + data_size);
```
nit: align object_size before adding it to data_size?

brian-pane replied:
`::operator new` guarantees that its return value will be "suitably aligned to hold an object of any fundamental alignment": https://en.cppreference.com/w/cpp/memory/new/operator_new

jmarantz replied:
I know that ... I was thinking that adding object_size and data_size without aligning either might  result in an allocation size that assumed a packing that would not correspond to the structure offsets. But now I can't think of a scenario where two realistic sizes would combine into something that wouldn't be aligned when passed to new(). So I guess this is OK.

Sorry I'm just being paranoid :)

brian-pane commented:
I'm trying to repro that error in the fuzz tests locally now.

brian-pane commented on test/common/buffer/buffer_fuzz_test.cc:
```diff
@@ -274,7 +274,7 @@ uint32_t bufferAction(Context& ctxt, char insert_value, uint32_t max_alloc, Buff
     const uint64_t slices_used = target_buffer.getRawSlices(slices, 1);
     if (linearize_size > 0) {
-      FUZZ_ASSERT(slices_used == 1);
+      FUZZ_ASSERT(slices_used >= 1);
```
Some context on this change: the buffer API specifies that `linearize(n)` will linearize at most the first `n` bytes of the buffer. The libevent buffer implementation happens to linearize the entire buffer; but the new buffer implementation stops at the next slice boundary after `n` bytes. The next line after this verifies that the first `n` bytes have indeed been linearized.

alyssawilk replied:
Just to sanity check, have we verfied that the in-envoy code sites either linearize(full buffer) or lineraize(n)/write(n) i.e. this change will end up being a no-op?

brian-pane replied:
Anyway, most of the call sites in Envoy use a third form instead: "linearize one field." But this is also safe with both the old and new implementations (and with any future implementation that satisfies the semantics defined in `buffer.h`) :
```
void* mem = buffer.linearize(n);
// read/interpret the n-byte field value
```
[Here](https://gist.github.com/brian-pane/07aa0a09b33b62dccad065aaf825d9a2) are all the `linearize` calls in Envoy. From that list, every call site uses one of those three safe patterns:
* All the call sites in mongo_proxy safely use the "linearize one field" pattern.
* dubbo_proxy provides a buffer wrapper class whose `linearize` method that maintains an internal offset within the buffer. That method itself looks correct, as long as its callers only rely on the documented `linearize` semantics.
* The http, mysql, and thrift filters safely use the "linearize one field" pattern.
* `source/extensions/transport_sockets/tls/ssl_socket.cc` uses the "linearize(n)/write(n)" pattern.
* `source/extensions/transport_sockets/alts/tsi_handshaker.cc` uses the "linearize(n)/write(n)" pattern. (The write is encapsulated in a gRPC library function, but the Envoy code passes the right value of "n" to it.)
* `source/extensions/transport_sockets/alts/tsi_frame_protector.cc` uses "linearize(buffer.length())"

So all the call sites in Envoy do the right thing. That's good, because even independent of this PR, the old implementation can't be assumed to linearize the entire buffer either. The underlying libevent function it uses, `evbuffer_pullup`, is only documented as linearizing the first `n` bytes. It currently happens to linearize the whole buffer in some test cases where `n` is close to the buffer length, but that's not guaranteed in the general case or in future releases of libevent.

brian-pane commented:
/retest

brian-pane commented:
/retest

brian-pane commented:
/retest

brian-pane commented:
The `ci/circleci: coverage` test keeps failing with a timeout. I don't think it's due to the changes in this PR, but I can't find enough diagnostic clues in the CircleCI output to verify the cause.

lizan commented:
@brian-pane I can reproduce this locally, just run `test/run_envoy_bazel_coverage.sh`. Likely due to some singleton related thing as all tests linked into one binary in coverage.

brian-pane commented:
@lizan how are you able to run `test/run_envoy_bazel_coverage.sh` locally? It fails before it can even start compiling:
```
source/exe/BUILD:26:1: Illegal ambiguous match on configurable attribute "linkopts" in //source/exe:envoy-static:
//bazel:coverage_build
@bazel_tools//tools/osx:darwin
Multiple matches are not allowed unless one is unambiguously more specialized.
```

jmarantz commented:
@brian-pane are you developing on Mac? I only ask because of `@bazel_tools//tools/osx:darwin` from your quoted output.

FWIW that script test/run_envoy_bazel_coverage.sh works out-of-the-box for me too.

Could this be related to whatever mechanism is used to latch whether you are using the old or new impl, and its impact on other tests that are all linked into the same binary when running with coverage?

brian-pane commented:
Yes, that's on Mac. I'll try switching to Linux. It's possible that the timeout issue is related to the runtime switch that selects between the old and new buffer impls (mixing the two isn't supported).

brian-pane commented:
Confirmed: the problem is that combined coverage test tries to use both types of buffers in the same process. The safeguard code in the buffer impl catches this via an assert, and the assertion signal handler hangs forever on a read when trying to get the symbols from the executable.

```
#0  0x00007ff8d299a384 in __libc_read (fd=0, buf=0x7ff852d45840, nbytes=1024)
    at ../sysdeps/unix/sysv/linux/read.c:27
#1  0x0000559abfa96ae2 in absl::debugging_internal::ReadPersistent (fd=0, buf=0x7ff852d45840, 
    count=1024) at external/com_google_absl/absl/debugging/symbolize_elf.inc:399
#2  0x0000559abfa96d10 in absl::debugging_internal::ReadFromOffset (fd=0, buf=0x7ff852d45840, 
    count=1024, offset=1920413272) at external/com_google_absl/absl/debugging/symbolize_elf.inc:424
#3  0x0000559abfa96f64 in absl::debugging_internal::GetSectionHeaderByType (fd=0, sh_num=47, 
    sh_offset=1920413272, type=2, out=0x7ff8d31ee210, tmp_buf=0x7ff852d45840 "\022\004", 
    tmp_buf_size=1024) at external/com_google_absl/absl/debugging/symbolize_elf.inc:466
#4  0x0000559abfa982fe in absl::debugging_internal::(anonymous namespace)::Symbolizer::GetSymbolFromObjectFile (this=0x7ff852d45020, obj=..., pc=
    0x559abf9ea5a6 <Envoy::SignalAction::sigHandler(int, siginfo_t*, void*)+202>, 
    relocation=94122865115136, out=0x7ff852d45032 "", out_size=2048, 
    tmp_buf=0x7ff852d45840 "\022\004", tmp_buf_size=1024)
    at external/com_google_absl/absl/debugging/symbolize_elf.inc:762
#5  0x0000559abfa9a993 in absl::debugging_internal::(anonymous namespace)::Symbolizer::GetSymbol (
    this=0x7ff852d45020, 
    pc=0x559abf9ea5a6 <Envoy::SignalAction::sigHandler(int, siginfo_t*, void*)+202>)
    at external/com_google_absl/absl/debugging/symbolize_elf.inc:1295
#6  0x0000559abfa9b59c in absl::Symbolize (
    pc=0x559abf9ea5a6 <Envoy::SignalAction::sigHandler(int, siginfo_t*, void*)+202>, 
    out=0x7ff8d31ee3e0 "\020\344\036\323\370\177", out_size=1024)
    at external/com_google_absl/absl/debugging/symbolize_elf.inc:1453
#7  0x0000559abb110a76 in Envoy::BackwardsTrace::visitTrace(std::function<void (int, char const*, void*)> const&) (this=0x7ff8d31ee890, visitor=...)
    at bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:107
#8  0x0000559abb11093f in Envoy::BackwardsTrace::logTrace (this=0x7ff8d31ee890)
    at bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:71
#9  0x0000559abf9ea62e in Envoy::SignalAction::sigHandler (sig=6, info=0x7ff8d31eebf0, 
    context=0x7ff8d31eeac0) at source/exe/signal_action.cc:20
#10 <signal handler called>
#11 __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#12 0x00007ff8d1c99801 in __GI_abort () at abort.c:79
#13 0x0000559abf99478b in Envoy::Buffer::OwnedImpl::useOldImpl (use_old_impl=true)
    at source/common/buffer/buffer_impl.cc:693
#14 0x0000559abdb0430c in Envoy::Server::InstanceImpl::initialize (this=0x559ad181c800, options=
    ..., local_address=
    std::shared_ptr<const Envoy::Network::Address::Instance> (use count 3, weak count 0) = {...}, 
    component_factory=...) at source/server/server.cc:230
```

htuch commented:
@brian-pane thanks, I think https://github.com/envoyproxy/envoy/pull/5441/commits/cfd9d4a068e30f8fa29e7474496fe2e493b8cd01 will make integration with the fuzz infra and supporting multiple implementations simpler as well.

jmarantz commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -217,21 +638,69 @@ OwnedImpl::OwnedImpl(const Instance& data) : OwnedImpl() { add(data); }
+  appendSliceForTest(data.data(), data.size());
 }
 
+void OwnedImpl::useOldImpl(bool use_old_impl) { use_old_impl_ = use_old_impl; }
```
To ensure that this only gets called during the quiescent state between tests, there's a simple way now.

Have a static bool committed_ alongside use_old_impl_. In this method do:
```
void OwnedImpl::useOldImpl(bool use_old_impl) {
  ASSERT(!committed_);
  use_old_impl_ = use_old_impl;
}
```

then you can write:
```
void OwnedImpl::clearCommitted() { committed_ = false; }
```
which you can call from here:

https://github.com/envoyproxy/envoy/blob/741df7aa13ddb15acbc77468579f0c6b58542839/test/test_listener.cc#L15

That will allow switching implementations between test methods.

Actually I'm surprised this didn't bite you before, aren't you using TEST_P to test with both mechanisms in the same binary?

brian-pane replied:
Up until [cfd9d4a](https://github.com/envoyproxy/envoy/commit/cfd9d4a068e30f8fa29e7474496fe2e493b8cd01), the buffer unit tests used a separate, test-only method that enabled the implementation to be changed.

Starting with that change, it's valid to switch back and forth between the implementations in the same process. The one remaining restriction is that two buffers can only interact (e.g., with `buffer1.move(buffer2)`) if they use the same implementation.

That enables the coverage test to progress further, but eventually it still hangs with an exception somewhere. It will take some more digging to find out where that's happening, as the hang happens in some test code that has caught the exception and is trying to log it.

jmarantz replied:
Right; what I'm trying to suggest is that it still be invalid to switch between implementations within a process, allowing only for switching after a test method.

That should cover the case of the coverage tests, and also fuzz tests, both of which will work with the test listener as described.

I think we probably want to RELEASE_ASSERT, actually, that we only call useOldImpl once per process/test-method.

My thought was that like catching things at this level would be a stronger guarantee of the invariant that only one impl be alive at a time.

Of course you still have to finish your debugging :) There's a chance this suggestion may catch it, but it's probably something else.

jmarantz commented on test/common/buffer/utility.h:
```diff
@@ -10,6 +10,36 @@ namespace Envoy {
+class BufferImplementationParamTest : public testing::TestWithParam<BufferImplementation> {
+protected:
+  BufferImplementationParamTest() {
+    OwnedImpl::useOldImpl(GetParam() == BufferImplementation::Old);
```
independent of coverage testing, how did this even work with just this one file? Wouldn't this have hit your assert?

Does this suggest you were not successfully testing both impls with this structure? It seems you should've been hitting your assert within that one test binary.

brian-pane replied:
Prior to the changes I added yesterday, the test code called a different method, `useOldImplForTest`, which didn't do the assert.

The test code has a call to `verifyImplementation(buffer);` in every test case to ensure that it's using the expected implementation.

brian-pane replied:
Now that the code allows switching buffer implementations in non-test code, `useOldImplForTest` is no more, and the unit tests just use the same `useOldImpl` that the production code uses.

jmarantz replied:
OK, this is fine. My suggestion was based around a theory that we should have safeguards against switching modes except at the start of a binary and between test-methods, but it's really a belt and suspenders argument and as long as you don't actually call it after startup in production it should be fine.

brian-pane commented on docs/root/intro/version_history.rst:
```diff
@@ -31,6 +31,7 @@ Version history
 * mysql: added a MySQL proxy filter that is capable of parsing SQL queries over MySQL wire protocol. Refer to ::ref:`MySQL proxy<config_network_filters_mysql_proxy>` for more details.
 * http: added :ref:`max request headers size <envoy_api_field_config.filter.network.http_connection_manager.v2.HttpConnectionManager.max_request_headers_kb>`. The default behaviour is unchanged.
 * http: added modifyDecodingBuffer/modifyEncodingBuffer to allow modifying the buffered request/response data.
+* performance: new buffer implementation (disabled by default; to test it, add "--use-libevent-buffers 0" to the command-line arguments when starting Envoy).
```
@rgs1 would you be interested in testing this with production traffic in your environment? It requires a command-line flag to turn it on.

mattklein123 replied:
I think the plan here should be:
1) Let's ship this PR
2) Let's set up fuzzing via @htuch. I agree with @htuch that if fuzzing doesn't uncover any issues I will feel pretty confident.
3) Once any fuzz issues are fixed, let's see if we can get a couple of production users to smoke test (envoy-dev/envoy-users email maybe).
4) Then remove the CLI flag, flip to default true, and config guard with the new stuff @alyssawilk has in https://github.com/envoyproxy/envoy/pull/6134

WDYT?

alyssawilk replied:
Having finally at least skimmed through this whole PR, I'm not sure using a reloadable flag is appropriate.  Many of the functions only work when comparing old buffer to old buffer and new to new, so I think we'd either have to do some plumbing to make sure the state was consistent throughout all buffers on a given connection (which might not even work then given muxed client connections can share a different muxed upstream connection) or we'd have to implement restart flags as well.

Rather one can latch the value of the runtime once where it's currently set, but then it's not actually doing new-style buffers until after a restart.  internally we make clear the differences between reloadable and restart flags so you know the difference between when you push things and when they take effect.  Alternately we could use the new config guard code it'd just be fairly misleading and we'd want to communicate it doesn't work as advertised in this case.

brian-pane replied:
This PR used to latch the value just once per process, except for a test override method. But the problem with that was that it broke the coverage test, which expected to be able to concatenate together a lot of tests that originally were designed to be separate programs with their own control over the choice of buffer implementation.

brian-pane commented on source/common/http/http2/metadata_decoder.cc:
```diff
@@ -54,22 +54,23 @@ bool MetadataDecoder::decodeMetadataPayloadUsingNghttp2(bool end_metadata) {
     // is_end indicates if the data in slice is the last data in the current
```
The reason for the changes in the source/common/http/http2/metadata_decoder.cc is that, while iterating on this PR, I changed the `RawSlice` declaration to make the `mem_` field immutable by code outside the buffer implementation. The one place that modified that field was the metadata encoder, so I changed metadata_decoder.cc to not require the field to be mutable.

I later reverted the having `RawSlice.mem_` be mutable, so the metadata_decoder.cc change here isn't strictly necessary any more. I do like the idea of keeping this change, though, because there are cases where it's unsafe to modify that `mem_` field, and the change here removes an example that people might otherwise copy and unwittingly use in an unsafe context.

mattklein123 replied:
I agree we should change this, but do you mind moving into a different PR?

brian-pane replied:
Ok, I've removed it from this PR.

mattklein123 commented:
@brian-pane also needs a master merge.

lizan commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -217,21 +638,69 @@ OwnedImpl::OwnedImpl(const Instance& data) : OwnedImpl() { add(data); }
-  size_t len = 0;
-  for (const RawSlice& slice : slices) {
-    len += slice.len_;
+  if (old_impl_) {
```
Does this method need if? Seems same and not depends on old/new impl?

brian-pane replied:
Fixed in the latest update

brian-pane commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -44,171 +62,574 @@ void OwnedImpl::add(const Instance& data) {
+    // Next, scan forward and attempt to match the slices against iovecs.
+    uint64_t num_slices_committed = 0;
+    while (num_slices_committed < num_iovecs) {
+      if (slices_[slice_index]->commit(iovecs[num_slices_committed])) {
```
Right, the new implementation keeps the ASSERT because the failure to commit all of the supplied iovecs is probably a sign of a major bug. The two failure scenarios I can think of are:
* The caller did the right thing and tried to properly commit iovecs that it got from `buffer.reserve()`, but the commit didn't work due to some internal bug in the buffer code.
* Or the caller did the wrong thing and tried to `buffer.commit()` something that didn't come from that `buffer`.

htuch replied:
FWIW, having as many ASSERTs as possible for consistency and debug checks has some history of paying off with the buffers + fuzzers. Please keep adding!

brian-pane commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -44,171 +62,574 @@ void OwnedImpl::add(const Instance& data) {
+      if (i < out_size) {
+        out[i].mem_ = slice->data();
+        out[i].len_ = slice->dataSize();
+      }
```
We have to loop through the all the slices because of the pre-existing semantics defined for the method in [include/envoy/buffer/buffer.h](https://github.com/envoyproxy/envoy/blob/master/include/envoy/buffer/buffer.h#L131)
```
   * @return the actual number of slices needed, which may be greater than out_size. Passing
   *         nullptr for out and 0 for out_size will just return the size of the array needed
   *         to capture all of the slice data.
```
I'll add a comment in the impl to explain why it does the counterintuitive thing.

brian-pane commented on source/common/buffer/buffer_impl.cc:
Okay, I'll add a RELEASE_ASSERT that covers both the old and new impls.

brian-pane commented on source/common/buffer/buffer_impl.cc:
Now that you mention it, the dynamic cast is redundant now, because the `ASSERT(isSameBufferImpl(rhs))` at the start of the method already encapsulates the same check (and does additional, more thorough checks). I'll remove the dynamic_cast here.

brian-pane commented on source/common/buffer/buffer_impl.cc:
Fixed in the latest push

brian-pane commented on source/common/buffer/buffer_impl.cc:
```diff
@@ -12,25 +12,43 @@
+  } else {
+    const char* src = static_cast<const char*>(data);
+    bool new_slice_needed = slices_.empty();
+    while (size != 0) {
```
There are several places in the buffer implementation where the length could overflow (e.g., adding buffer A to buffer B if they both contain 2**63 bytes), so I just pushed an update that wraps all the buffer length arithmetic operations in an overflow/underflow detector that clang and g++ provide.

brian-pane replied:
Based on the CircleCI errors, the builtin overflow/underflow detector seems to have some portability problems. I'll have to write one from scratch.

brian-pane commented:
I've addressed all the review comments, so this is ready for re-review.

mattklein123 commented:
@jmarantz @alyssawilk @envoyproxy/maintainers any additional thoughts on this one? I think we should ship and iterate.

alyssawilk commented:
+1.

I'm for checking this in, fuzzing the heck out of it, (ideally but optionally) getting a post-fuzzing review pass from someone security minded, canarying somewhere, then defaulting true.

jmarantz commented:
+1 -- having a security-minded review would be good; do we have someone on the project that would be suitable?

mattklein123 commented:
OK let's merge and iterate. We can discuss next steps once initial fuzzing is complete and any bugs are fixed. @brian-pane thank you for all your hard work here and being such a trooper with reviews.
